File: backend/src/main/java/com/NBE_4_5_2/Team5/global/security/CustomAuthorizationRequestResolver.java
package com.NBE_4_5_2.Team5.global.security;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpSession;
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizationRequestResolver;
import org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestResolver;
import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;
import org.springframework.stereotype.Component;

@Component
public class CustomAuthorizationRequestResolver implements OAuth2AuthorizationRequestResolver {

    private final OAuth2AuthorizationRequestResolver defaultResolver;

    public CustomAuthorizationRequestResolver(ClientRegistrationRepository clientRegistrationRepository) {
        this.defaultResolver = new DefaultOAuth2AuthorizationRequestResolver(clientRegistrationRepository, "/oauth2/authorization");
    }

    @Override
    public OAuth2AuthorizationRequest resolve(HttpServletRequest request) {
        OAuth2AuthorizationRequest authorizationRequest = defaultResolver.resolve(request);

        return customizeAuthorizationRequest(request, authorizationRequest);
    }

    @Override
    public OAuth2AuthorizationRequest resolve(HttpServletRequest request, String clientRegistrationId) {
        OAuth2AuthorizationRequest authorizationRequest = defaultResolver.resolve(request, clientRegistrationId);

        return customizeAuthorizationRequest(request, authorizationRequest);
    }

    private OAuth2AuthorizationRequest customizeAuthorizationRequest(HttpServletRequest request,
                                                                     OAuth2AuthorizationRequest authorizationRequest) {
        if (authorizationRequest == null) {
            return null;
        }

        String customParam = request.getParameter("redirectUrl");

        if (customParam != null) {
            HttpSession session = request.getSession();
            session.setAttribute("redirectUrl", customParam);
        }

        return OAuth2AuthorizationRequest.from(authorizationRequest).build();
    }
}

File: backend/src/main/java/com/NBE_4_5_2/Team5/global/security/CustomAuthorizationRequestResolver.kt
package com.NBE_4_5_2.Team5.global.security

import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository
import org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizationRequestResolver
import org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestResolver
import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest
import org.springframework.stereotype.Component

@Component
class CustomAuthorizationRequestResolver(
    clientRegistrationRepository: ClientRegistrationRepository
) : OAuth2AuthorizationRequestResolver {

    private val defaultResolver =
        DefaultOAuth2AuthorizationRequestResolver(
            clientRegistrationRepository,
            "/oauth2/authorization"
        )

    override fun resolve(request: HttpServletRequest): OAuth2AuthorizationRequest? =
        defaultResolver.resolve(request)
            ?.let { customize(request, it) }

    override fun resolve(
        request: HttpServletRequest,
        clientRegistrationId: String
    ): OAuth2AuthorizationRequest? =
        defaultResolver.resolve(request, clientRegistrationId)
            ?.let { customize(request, it) }

    private fun customize(
        request: HttpServletRequest,
        req: OAuth2AuthorizationRequest
    ): OAuth2AuthorizationRequest {
        request.getParameter("redirectUrl")
            ?.also { request.session.setAttribute("redirectUrl", it) }
        return OAuth2AuthorizationRequest.from(req).build()
    }
}
